src/store/modules/accounts/hdWallet.js

```diff
@@ -30,16 +38,44 @@ export default {
       return aeSdk.api.getAccountByPubkey(address).then(() => true, () => false);
     },
     async discover({ rootGetters: { wallet }, dispatch }) {
-      const numberOfAccounts = await ProtocolAdapterFactory
-        .getAdapter(PROTOCOL_AETERNITY) // Discover only aeternity accounts for now
-        .discoverAccounts(wallet);
-
-      for (let i = 0; i < numberOfAccounts; i += 1) {
-        dispatch('create', {
-          isRestored: true,
-          protocol: PROTOCOL_AETERNITY,
-        });
-      }
+      const accountsToRecover = await Promise.all(PROTOCOLS.map(async (protocol) => {
+        let lastNotEmptyIdx = 0;
+        let lastIndex = 0;
+        let isAccountUsedArray = [];
+        const getProtocolAdapter = await ProtocolAdapterFactory.getAdapter(protocol);
+        do {
+          try {
+            lastNotEmptyIdx = isAccountUsedArray.lastIndexOf(true) + lastIndex;
+            lastIndex += isAccountUsedArray.length;
+            // eslint-disable-next-line no-await-in-loop
+            isAccountUsedArray = await Promise.all(
+              Array.from(
+                Array(ADDRESS_GAP_LIMIT + lastNotEmptyIdx - lastIndex + 1).keys(),
+                // eslint-disable-next-line no-loop-func
+                (x) => x + lastIndex,
+              )
+                .map((index) => getProtocolAdapter.isAccountUsed(
+                  getProtocolAdapter.getHdWalletAccountFromMnemonicSeed(wallet, index).address,
+                )),
+            );
+          } catch (e) {
+            break;
+          }
+        } while (!(
+          isAccountUsedArray.lastIndexOf(true) === -1
+            || isAccountUsedArray.filter((isAccountUsed) => !isAccountUsed)
+              .length === ADDRESS_GAP_LIMIT
+        ));
+        return { protocol, lastNotEmptyIdx };
+      }));
+      accountsToRecover.forEach(({ protocol, lastNotEmptyIdx }) => {
+        for (let i = protocol === PROTOCOL_AETERNITY ? 1 : 0; i <= lastNotEmptyIdx; i += 1) {
+          dispatch('create', {
+            isRestored: true,
+            protocol,
+          });
+        }
+      });
```

Comment:
Please leave a code comment on the account discovery algorithm. The approach looks complicated. This could be simplified by breaking the single-function into multiple functions.

Reply:
Standard algorithm for Account discovery https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account-discovery

Reply:
This is exactly what we are doing (with the fact that we are using only `account`, without `addresses`).

Reply:
Actually there is an explanation above `ADDRESS_GAP_LIMIT` variable, that explains the logic behind account discovery.

Reply:
Sure if you think that describes the logic, No problem then.